pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/parser/RuntimeASTTransformer.groovy

```diff
@@ -0,0 +1,759 @@
+        } else if (s instanceof TryCatchStatement) {
+            // Translate the try and finally statements, as well as any catch statements
+            TryCatchStatement t = (TryCatchStatement) s
+            TryCatchStatement newTry = new TryCatchStatement(translateClosureStatement(targetVar, t.tryStatement, keys),
+                translateClosureStatement(targetVar, t.finallyStatement, keys))
+            t.catchStatements.each { c ->
+                newTry.addCatch((CatchStatement)translateClosureStatement(targetVar, c, keys))
+            }
+            return newTry
+        } else if (s instanceof CatchStatement) {
+            // Translate the nested statement
+            return catchS(s.variable, translateClosureStatement(targetVar, s.code, keys))
+        } else {
+            throw new IllegalArgumentException("Got an unexpected " + s.getClass() + " in environment, please " +
+                "report at issues.jenkins-ci.org")
+        }
+    }
+
+    /**
+     * Generates the method call for fetching the closure for a given environment key and calling it.
+     */
+    private MethodCallExpression environmentValueGetterCall(String name) {
+        return callX(callThisX("getClosure", constX(name)), "call")
+    }
+
+    /**
+     * Generates the AST (to be CPS-transformed) for instantiating {@link Libraries}.
+     *
+     * @param original The parsed AST model
+     * @return The AST for a constructor call for {@link Libraries}, or the constant null expression if the original
+     * cannot be transformed.
+     */
+    Expression transformLibraries(@CheckForNull ModelASTLibraries original) {
+        if (isGroovyAST(original) && !original.libs.isEmpty()) {
+            ListExpression listArg = new ListExpression()
+            original.libs.each { l ->
+                if (l.sourceLocation instanceof Expression) {
+                    listArg.addExpression((Expression)l.sourceLocation)
+                }
+            }
+            return ctorX(ClassHelper.make(Libraries.class), args(listArg))
+        }
+
+        return constX(null)
+    }
+
+    /**
+     * Generates the AST (to be CPS-transformed) for instantiating {@link Options}.
+     *
+     * @param original The parsed AST model
+     * @return The AST for a constructor call for {@link Options}, or the constant null expression if the original
+     * cannot be transformed.
+     */
+    Expression transformOptions(@CheckForNull ModelASTOptions original) {
+        if (isGroovyAST(original) && !original.options.isEmpty()) {
+            List<ModelASTOption> jobProps = new ArrayList<>()
+            List<ModelASTOption> options = new ArrayList<>()
+            List<ModelASTOption> wrappers = new ArrayList<>()
+
+            SymbolLookup symbolLookup = SymbolLookup.get()
+
+            original.options.each { o ->
+                if (symbolLookup.findDescriptor(JobProperty.class, o.name) != null) {
```

Comment:
Yeah, that's done in `ModelValidatorImpl` - we never get here if there are invalid `JobProperty`s etc.